Add vscode remote

michelav opened the pull request:
This provides means for using vscode remote development extension. I'm using the default node image provided by vscode \ ms (no Dockerfile needed) . Besides the configuration file (.devcontainer.json), the changes are:

- Commented out all networks at harhat.config.js. This way `npm run build` works out of the box with hardhat local network and there is no complain about misconfigurations;
- Updated README with information about vscode remote extension and also explaining how to configure other networks besides hardhat's built-in.

I've tested vscode remote setup at Linux and Win11. Both worked OK.

michelav committed:
Add vscode remote dev container support

michelav committed:
Adjust configuration files

michelav committed:
Update setup and development information

haltakov commented on README.md:
```diff
@@ -2,25 +2,60 @@
 
 ## Develop
 
+Smart contracts and NFTs demand an account and a blockchain network to be deployed into. Besides that, it's important to set an
```
Thanks for the detailed docs. We already have some detailed description of how to set up the environment here, but I think we forgot to link it in the README... 

https://docs.underfitted.io/incubator/bnomial/web3/environment-setup

michelav replied:
Yeah... saw that. The doc there is much more detailed. I put this here because I felt like something was missing and a little bit of explanation would help when building the stuff. So, should I remove and just mention the link to the docs?

michelav replied:
We could have a line in the beginning of Set Up section like this:

> There is a detailed description of Bnomial project and some guidelines regarding its set up and development. You can find them in this [link](https://docs.underfitted.io/incubator/bnomial/web3/environment-setup).

What do you think?

haltakov replied:
Yes, I like that. Let's keep the rest of the info you wrote here. We just need to be aware that both sources may get out of sync at some point.

haltakov replied:
Added this text further down.

haltakov commented on hardhat.config.js:
```diff
@@ -10,14 +10,15 @@ module.exports = {
-            url: POLYGON_API_URL,
-            accounts: [`0x${PRIVATE_KEY}`],
-        },
+        // Uncomment according to the network you wanna use
```
Hmmm, why do we need to comment out these lines? Is it because of the missing `.env` file? If so, maybe we can have a default one generated in the container like we do in the CI (see [here](https://github.com/underfitted/bnomial-web3/blob/7833cb76c7295f8ff524c16c761953a82a180aa3/.github/workflows/test.yml#L31))? Or maybe another way to work around this in the config file

michelav replied:
The way how it was if someone tried to run locally using the Hardhat dummy network (which demands no configuration at all) it would be necessary to set a fake private key (`npm run build` complains about values undefined and formats). It felt like old practices we had before things like [12-factor app](https://12factor.net). In fact, commenting them out isn't a good solution either.

If we want to use the ideas from 12-factor apps, we should change the configuration file for something like this:

````
...
networks: {
      hardhat: {},
      deploynet: {
          url: https://myaccessprovider.com/api/v2/${API_KEY},
          accounts: [${PRIVATE_KEY}],
      },
},
etherscan: {
      apiKey: ${POLYSCAN_KEY},
},
...
````

We would avoid grouping configuration by deployment perimeter and it would allow our CI engine to use the same deployment asset (the docker image) and promote it through a pipeline without touching the config file (e.g. the container could run and deploy in Mumbai, Polygon or a new private network we want just by changing the container ENV variables).

michelav replied:
I guess that these changes wouldn't fit well in this branch. We should keep the way it was before my commits and discuss the configuration file (if possible) in another branch.

haltakov replied:
I'll look into that today. There should be a way to set some defaults so to not block people without `.env` setup.

haltakov replied:
I like the idea of refactoring the config file, but as you say let's do it on another branch. In the meantime I think I have a nice workaround :)

haltakov committed:
Add defaults for env variables

This allows users to run all local commands without creating an .env
file by setting default values for all env variables.

haltakov committed:
Add link to the VSCode Remote Container extension

haltakov committed:
Add link to environment setup in the official docs

haltakov commented on hardhat.config.js:
```diff
@@ -4,7 +4,10 @@ require("hardhat-gas-reporter");
 require("dotenv").config();
 
-const { MUMBAI_API_URL, POLYGON_API_URL, PRIVATE_KEY, POLYSCAN_KEY } = process.env;
+const MUMBAI_API_URL = process.env.MUMBAI_API_URL || "";
```
I now added defaults for all env variables so we don't need dummy .env files anymore.

haltakov committed:
Merge branch 'main' into add_vscode_remote
